pytorch_lightning/plugins/layer_sync.py

```diff
@@ -0,0 +1,98 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+
+import torch
+from torch.nn import Module
+
+
+class LayerSync(ABC):
```

Comment:
Have you considered making this a contextmanager even if we don't use it with a simple `with` statement?

Reply:
it requires the model as input. i don't see how a context manager could possibly access the model(s) instances under it.

Reply:
Something like this?

```python
class LayerSync:
    def __init__(self, model):
        self.model = model  # weakref.proxy?

    def __enter__(self) -> Module:
        pass

    def __exit__(self, exc_type, exc_value, exc_traceback) -> Module:
        pass

    def apply(self) -> Module:
        return self.__enter__()

    def revert(self) -> Module:
        return self.__exit__(None, None, None)
```

Reply:
I still don't see this. It has no effect on the code under the context manager. What is the meaning of this?

```py
with LayerSync(model):
    model(x)
    model2(x)
```

Let's not do this. Even if it somehow made sense, nobody can't take advantage of this "feature". Also, we don't want people using plugins outside the trainer. That will only limit us later when we want to develop the plugin further.